### ISSUE TITLE
Phase 2.3: Confidence Scoring and Calibration System

### ISSUE THREAD

**Issue by justin4957** [justin4957/loft#27]
## Phase 2: LLM Logic Generation + Validation
**Weeks 6-9 deliverable**

### Objective
Implement comprehensive confidence scoring system that calibrates and tracks confidence at every level: rule generation, validation stages, and empirical performance. Confidence scores must be meaningful predictors of actual accuracy.

### Background
Confidence scores are critical for:
1. **Gating decisions**: Rules below threshold rejected/flagged
2. **Stratified modification**: Higher layers require higher confidence
3. **Self-reflexivity**: System must know what it doesn't know
4. **Risk assessment**: Identify uncertain areas needing human review

From CLAUDE.md: "Maintaining Epistemic Humility - The system must know what it doesn't know."

#### 1. Multi-Source Confidence Aggregation

Confidence comes from multiple sources:
- **Generation Confidence**: LLM's self-assessed confidence
- **Validation Confidence**: Pipeline validation results
- **Empirical Confidence**: Performance on test cases
- **Consensus Confidence**: Agreement among multiple LLMs
- **Historical Confidence**: Past performance of similar rules

**Aggregation Strategy**:
```python
class ConfidenceAggregator:
    """Aggregate confidence from multiple sources."""

    def __init__(self, weights: Optional[Dict[str, float]] = None):
        # Default weights
        self.weights = weights or {
            "generation": 0.15,
            "syntax": 0.10,
            "semantic": 0.15,
            "empirical": 0.40,  # Highest weight - actual performance
            "consensus": 0.20,
        }

        # Ensure weights sum to 1.0
        total = sum(self.weights.values())
        self.weights = {k: v/total for k, v in self.weights.items()}

    def aggregate(
        self,
        generation_confidence: float,
        syntax_valid: bool,
        semantic_valid: bool,
        empirical_accuracy: float,
        consensus_strength: float
    ) -> AggregatedConfidence:
        """
        Compute weighted confidence score.

        Args:
            generation_confidence: LLM's confidence in generated rule
            syntax_valid: Did rule pass syntax validation
            semantic_valid: Did rule pass semantic validation
            empirical_accuracy: Accuracy on test cases
            consensus_strength: Agreement among voter LLMs

        Returns:
            AggregatedConfidence with score and breakdown
        """

        components = {
            "generation": generation_confidence,
            "syntax": 1.0 if syntax_valid else 0.0,
            "semantic": 1.0 if semantic_valid else 0.0,
            "empirical": empirical_accuracy,
            "consensus": consensus_strength,
        }

        # Weighted sum
        aggregate_score = sum(
            self.weights[source] * score
            for source, score in components.items()
        )

        # Compute variance to measure agreement
        variance = self._compute_variance(components)

        return AggregatedConfidence(
            score=aggregate_score,
            components=components,
            weights=self.weights,
            variance=variance,
            is_reliable=variance < 0.1  # Low variance = high agreement
        )

    def _compute_variance(self, components: Dict[str, float]) -> float:
        """Measure disagreement among confidence sources."""
        scores = list(components.values())
        mean = sum(scores) / len(scores)
        return sum((s - mean) ** 2 for s in scores) / len(scores)

class AggregatedConfidence(BaseModel):
    """Aggregated confidence from multiple sources."""
    score: float = Field(ge=0.0, le=1.0)
    components: Dict[str, float]
    weights: Dict[str, float]
    variance: float
    is_reliable: bool
    timestamp: datetime = Field(default_factory=datetime.now)

    def explanation(self) -> str:
        """Human-readable confidence explanation."""
        lines = [f"Overall Confidence: {self.score:.2f}"]
        lines.append("\nBreakdown:")
        for source, score in self.components.items():
            weight = self.weights[source]
            contribution = weight * score
            lines.append(
                f"  {source.title()}: {score:.2f} "
                f"(weight: {weight:.2f}, contribution: {contribution:.2f})"
            )
        lines.append(f"\nVariance: {self.variance:.3f} ({'reliable' if self.is_reliable else 'unreliable'})")
        return "\n".join(lines)
```

#### 2. Confidence Calibration

**Problem**: LLM-reported confidence often poorly calibrated with actual accuracy.

**Solution**: Empirical calibration using historical data.

```python
class ConfidenceCalibrator:
    """Calibrate confidence scores to match empirical accuracy."""

    def __init__(self):
        self.calibration_data: List[CalibrationPoint] = []
        self.calibration_function: Optional[Callable] = None

    def record(self, predicted_confidence: float, actual_accuracy: float, rule_id: str):
        """Record a calibration data point."""
        self.calibration_data.append(
            CalibrationPoint(
                predicted=predicted_confidence,
                actual=actual_accuracy,
                rule_id=rule_id
            )
        )

    def calibrate(self, method: str = "isotonic") -> CalibrationReport:
        """
        Learn calibration function from historical data.

        Methods:
        - "isotonic": Isotonic regression (monotonic, non-parametric)
        - "platt": Platt scaling (logistic regression)
        - "beta": Beta calibration
        """

        if len(self.calibration_data) < 10:
            logger.warning(f"Only {len(self.calibration_data)} calibration points, may be unreliable")

        predicted = [p.predicted for p in self.calibration_data]
        actual = [p.actual for p in self.calibration_data]

        if method == "isotonic":
            from sklearn.isotonic import IsotonicRegression
            calibrator = IsotonicRegression(out_of_bounds="clip")
            calibrator.fit(predicted, actual)
            self.calibration_function = calibrator.predict

        elif method == "platt":
            from sklearn.linear_model import LogisticRegression
            # Convert to classification problem (binning)
            calibrator = LogisticRegression()
            # Implementation details...

        # Evaluate calibration quality
        calibrated = [self.calibrate_score(p) for p in predicted]

        report = CalibrationReport(
            method=method,
            num_points=len(self.calibration_data),
            before_mse=self._mse(predicted, actual),
            after_mse=self._mse(calibrated, actual),
            before_ece=self._expected_calibration_error(predicted, actual),
            after_ece=self._expected_calibration_error(calibrated, actual),
        )

        logger.info(f"Calibration complete: {report}")
        return report

    def calibrate_score(self, raw_confidence: float) -> float:
        """Apply calibration to raw confidence."""
        if self.calibration_function is None:
            logger.warning("No calibration function, returning raw score")
            return raw_confidence

        return float(self.calibration_function([raw_confidence])[0])

    def _mse(self, predicted: List[float], actual: List[float]) -> float:
        """Mean squared error."""
        return sum((p - a) ** 2 for p, a in zip(predicted, actual)) / len(predicted)

    def _expected_calibration_error(
        self,
        predicted: List[float],
        actual: List[float],
        num_bins: int = 10
    ) -> float:
        """
        Expected Calibration Error (ECE).

        Bins predictions, compares average confidence to average accuracy in each bin.
        """
        bins = [[] for _ in range(num_bins)]

        for p, a in zip(predicted, actual):
            bin_idx = min(int(p * num_bins), num_bins - 1)
            bins[bin_idx].append((p, a))

        ece = 0.0
        total = len(predicted)

        for bin_points in bins:
            if not bin_points:
                continue

            bin_confidence = sum(p for p, _ in bin_points) / len(bin_points)
            bin_accuracy = sum(a for _, a in bin_points) / len(bin_points)
            bin_weight = len(bin_points) / total

            ece += bin_weight * abs(bin_confidence - bin_accuracy)

        return ece

class CalibrationPoint(BaseModel):
    predicted: float
    actual: float
    rule_id: str
    timestamp: datetime = Field(default_factory=datetime.now)

class CalibrationReport(BaseModel):
    method: str
    num_points: int
    before_mse: float
    after_mse: float
    before_ece: float  # Expected Calibration Error
    after_ece: float
    improvement: float = Field(default=0.0)

    def __init__(self, **data):
        super().__init__(**data)
        self.improvement = (self.before_ece - self.after_ece) / self.before_ece
```

#### 3. Confidence-Based Decision Making

**Stratified Confidence Thresholds** (from CLAUDE.md):

```python
class ConfidenceGate:
    """Gate decisions based on confidence thresholds."""

    # Default thresholds from CLAUDE.md
    THRESHOLDS = {
        "operational": 0.6,   # Frequent updates, lower risk
        "tactical": 0.8,      # Moderate impact
        "strategic": 0.9,     # High impact, requires high confidence
        "constitutional": 1.0,  # Requires human approval
    }

    def __init__(
        self,
        calibrator: Optional[ConfidenceCalibrator] = None,
        thresholds: Optional[Dict[str, float]] = None
    ):
        self.calibrator = calibrator
        self.thresholds = thresholds or self.THRESHOLDS.copy()

    def should_accept(
        self,
        confidence: AggregatedConfidence,
        target_layer: str,
        rule_impact: str = "medium"
    ) -> GateDecision:
        """
        Determine if rule should be accepted based on confidence.

        Args:
            confidence: Aggregated confidence score
            target_layer: Which stratification layer (operational/tactical/strategic)
            rule_impact: Estimated impact (low/medium/high)

        Returns:
            GateDecision with action and reasoning
        """

        # Calibrate if calibrator available
        calibrated_score = confidence.score
        if self.calibrator:
            calibrated_score = self.calibrator.calibrate_score(confidence.score)

        threshold = self.thresholds[target_layer]

        # Adjust threshold based on impact
        if rule_impact == "high":
            threshold += 0.1
        elif rule_impact == "low":
            threshold -= 0.05

        threshold = max(0.0, min(1.0, threshold))  # Clamp to [0,1]

        # Decision logic
        if calibrated_score >= threshold:
            if confidence.is_reliable:
                action = "accept"
                reasoning = f"Confidence {calibrated_score:.2f} meets threshold {threshold:.2f}"
            else:
                action = "flag_for_review"
                reasoning = f"Confidence sufficient but variance too high ({confidence.variance:.3f})"

        elif calibrated_score >= threshold - 0.1:
            action = "flag_for_review"
            reasoning = f"Confidence {calibrated_score:.2f} near threshold {threshold:.2f}"

        else:
            action = "reject"
            reasoning = f"Confidence {calibrated_score:.2f} below threshold {threshold:.2f}"

        return GateDecision(
            action=action,
            reasoning=reasoning,
            raw_confidence=confidence.score,
            calibrated_confidence=calibrated_score,
            threshold=threshold,
            target_layer=target_layer,
            rule_impact=rule_impact
        )

class GateDecision(BaseModel):
    action: Literal["accept", "reject", "flag_for_review"]
    reasoning: str
    raw_confidence: float
    calibrated_confidence: float
    threshold: float
    target_layer: str
    rule_impact: str
```

#### 4. Confidence Tracking and Analysis

```python
class ConfidenceTracker:
    """Track confidence scores over time for analysis."""

    def __init__(self, storage: Storage):
        self.storage = storage

    def track(self, rule_id: str, confidence: AggregatedConfidence):
        """Record confidence for a rule."""
        self.storage.store(f"confidence/{rule_id}", confidence.dict())

    def analyze_trends(self, window: int = 100) -> ConfidenceTrends:
        """Analyze confidence trends over recent rules."""

        recent = self.storage.get_recent(f"confidence/*", limit=window)

        scores = [c["score"] for c in recent]
        variances = [c["variance"] for c in recent]

        return ConfidenceTrends(
            mean_confidence=statistics.mean(scores),
            median_confidence=statistics.median(scores),
            std_confidence=statistics.stdev(scores) if len(scores) > 1 else 0.0,
            mean_variance=statistics.mean(variances),
            num_reliable=sum(1 for c in recent if c["is_reliable"]),
            num_total=len(recent),
            reliability_rate=sum(1 for c in recent if c["is_reliable"]) / len(recent)
        )

    def identify_underconfident_areas(self) -> List[str]:
        """Identify predicates/domains where confidence is consistently low."""
        # Group by predicate, find areas with low confidence
        # Returns list of predicates needing attention

    def identify_overconfident_areas(self) -> List[str]:
        """Identify where confidence exceeds actual performance."""
        # Compare predicted vs actual accuracy
        # Returns list of predicates with poor calibration

class ConfidenceTrends(BaseModel):
    mean_confidence: float
    median_confidence: float
    std_confidence: float
    mean_variance: float
    num_reliable: int
    num_total: int
    reliability_rate: float
```

### Deliverables

1. **Confidence Aggregation** (`loft/validation/confidence_aggregator.py`)
   - ConfidenceAggregator class
   - AggregatedConfidence schema

2. **Confidence Calibration** (`loft/validation/confidence_calibrator.py`)
   - ConfidenceCalibrator class
   - Isotonic/Platt/Beta calibration methods
   - CalibrationReport schema

3. **Confidence Gating** (`loft/validation/confidence_gate.py`)
   - ConfidenceGate class
   - Stratified threshold logic
   - GateDecision schema

4. **Confidence Tracking** (`loft/validation/confidence_tracker.py`)
   - ConfidenceTracker class
   - Trend analysis
   - Calibration monitoring

5. **Tests**
   - `tests/unit/test_confidence_aggregator.py`
   - `tests/unit/test_confidence_calibrator.py`
   - `tests/unit/test_confidence_gate.py`
   - `tests/integration/test_confidence_pipeline.py`

### Validation Criteria (from ROADMAP.md Phase 2)

- [ ] Confidence scores calibrated (±5% from ROADMAP.md core metrics)
- [ ] False positive rate <10% (high confidence, low actual accuracy)
- [ ] False negative rate <20% (low confidence, high actual accuracy)
- [ ] Expected Calibration Error (ECE) < 0.1
- [ ] Confidence variance predicts reliability
- [ ] Stratified thresholds enforced correctly

### MVP Success Metrics

- [ ] **Calibration Quality**: ECE < 0.1 after calibration
- [ ] **Prediction**: Confidence predicts accuracy within ±0.15
- [ ] **Reliability**: Low variance scores more accurate than high variance
- [ ] **Gating**: Correct accept/reject decisions >85% of time
- [ ] **Trend Detection**: Identifies under/over-confident areas

### Dependencies

- Phase 2.1: Rule Generation (#25) - provides generation confidence
- Phase 2.2: Validation Pipeline (#26) - provides validation confidence
- Phase 0.2: Metrics Framework (#2) ✅ - metrics tracking

### Related Issues

- Feeds into Phase 3: Safe Self-Modification (confidence-based rollback)
- Supports Phase 5: Meta-Reasoning (confidence about confidence)
- Enables human-in-the-loop (Phase 2.4)

### Testing Strategy

**Unit Tests**:
- Test aggregation with various component scores
- Test calibration with synthetic data (known relationships)
- Test gating logic with edge cases
- Verify threshold enforcement

**Integration Tests**:
- Calibrate on real rule generation data
- Validate calibration improves accuracy prediction
- Test full pipeline: generation → validation → aggregation → gating

**Property Tests**:
- Calibration never increases error
- Aggregated confidence always in [0,1]
- Higher confidence correlates with higher accuracy
- Variance correlates with prediction error

### Example Usage

```python
from loft.validation.confidence_aggregator import ConfidenceAggregator
from loft.validation.confidence_calibrator import ConfidenceCalibrator
from loft.validation.confidence_gate import ConfidenceGate

# Aggregate confidence from pipeline
aggregator = ConfidenceAggregator()
confidence = aggregator.aggregate(
    generation_confidence=0.87,
    syntax_valid=True,
    semantic_valid=True,
    empirical_accuracy=0.82,
    consensus_strength=0.90
)

print(confidence.explanation())
# Output:
# Overall Confidence: 0.84
# Breakdown:
#   Generation: 0.87 (weight: 0.15, contribution: 0.13)
#   Syntax: 1.00 (weight: 0.10, contribution: 0.10)
#   Semantic: 1.00 (weight: 0.15, contribution: 0.15)
#   Empirical: 0.82 (weight: 0.40, contribution: 0.33)
#   Consensus: 0.90 (weight: 0.20, contribution: 0.18)
# Variance: 0.045 (reliable)

# Calibrate confidence
calibrator = ConfidenceCalibrator()
# ... record many calibration points over time ...
calibration_report = calibrator.calibrate(method="isotonic")

calibrated_score = calibrator.calibrate_score(confidence.score)
print(f"Raw: {confidence.score:.2f} → Calibrated: {calibrated_score:.2f}")

# Gate decision
gate = ConfidenceGate(calibrator=calibrator)
decision = gate.should_accept(
    confidence=confidence,
    target_layer="tactical",
    rule_impact="medium"
)

print(f"Decision: {decision.action}")
print(f"Reasoning: {decision.reasoning}")
```

### Implementation Notes

- Calibration requires historical data - bootstrap with default calibration
- Recalibrate periodically (every 50-100 rules)
- Store calibration data for analysis
- Log all confidence computations with `@track_confidence` decorator
- Support multiple calibration methods (isotonic recommended)
- Provide visualization tools for calibration curves
- Alert when calibration degrades (ECE increases)

### Calibration Metrics Visualization

```python
def plot_calibration_curve(calibrator: ConfidenceCalibrator):
    """Plot predicted vs actual confidence."""
    import matplotlib.pyplot as plt

    predicted = [p.predicted for p in calibrator.calibration_data]
    actual = [p.actual for p in calibrator.calibration_data]

    # Before calibration
    plt.scatter(predicted, actual, alpha=0.5, label="Raw")

    # After calibration
    calibrated = [calibrator.calibrate_score(p) for p in predicted]
    plt.scatter(calibrated, actual, alpha=0.5, label="Calibrated")

    # Perfect calibration line
    plt.plot([0, 1], [0, 1], 'k--', label="Perfect Calibration")

    plt.xlabel("Predicted Confidence")
    plt.ylabel("Actual Accuracy")
    plt.legend()
    plt.title("Confidence Calibration")
    plt.show()
```

### Risks & Mitigations

**Risk**: Insufficient calibration data in early stages
**Mitigation**: Start with conservative thresholds, bootstrap with synthetic data

**Risk**: Calibration overfits to specific test cases
**Mitigation**: Use cross-validation, separate calibration and test sets

**Risk**: Confidence scores become overoptimistic over time
**Mitigation**: Continuous monitoring, recalibration, alert on ECE degradation

**Risk**: Weights in aggregation are suboptimal
**Mitigation**: Learn weights empirically, A/B test different weighting schemes
